src/Distribution/Server/Features/Vouch.hs

```diff
@@ -0,0 +1,218 @@
+  -- Make sure none of them are too recent.
+  when (length vouchersForVoucher >= requiredCountOfVouches && any (isWithinLastMonth now) vouchersForVoucher) $
+    throwError You'reTooNew
+  when (vouchee `Group.member` ugroup) $
+    throwError VoucheeAlreadyUploader
+  when (length existingVouchers >= 3) $
+    throwError AlreadySufficientlyVouched
+  when (voucher `elem` map fst existingVouchers) $
+    throwError YouAlreadyVouched
+  pure $
+    if length existingVouchers == requiredCountOfVouches - 1
+       then AddVouchComplete
+       else AddVouchIncomplete
+
+renderToLBS :: (UserId -> ServerPartE UserInfo) -> [(UserId, UTCTime)] -> ServerPartE TemplateAttr
+renderToLBS lookupUserInfo vouches = do
+  rendered <- traverse renderVouchers vouches
+  pure $
+    templateUnescaped "vouches" $
+      if null rendered
+         then LBS.pack "Nobody has vouched yet."
+         else LBS.intercalate mempty rendered
+  where
+  renderVouchers :: (UserId, UTCTime) -> ServerPartE LBS.ByteString
+  renderVouchers (uid, timestamp) = do
+    info <- lookupUserInfo uid
+    let UserName name = userName info
+        -- We don't need to show millisecond precision
+        -- So we truncate it off here
+        truncated = truncate $ utctDayTime timestamp
+        newUTCTime = timestamp {utctDayTime = secondsToDiffTime truncated}
+    pure . toUTF8LBS . prettyHtmlFragment . li . stringToHtml $ name <> " vouched on " <> formatShow iso8601Format newUTCTime
+
+initVouchFeature :: ServerEnv -> IO (UserFeature -> UploadFeature -> IO VouchFeature)
+initVouchFeature ServerEnv{serverStateDir, serverTemplatesDir, serverTemplatesMode} = do
+  vouchState <- vouchStateComponent serverStateDir
+  templates <- loadTemplates serverTemplatesMode [ serverTemplatesDir, serverTemplatesDir </> "Html"]
+                                                 ["vouch.html"]
+  vouchTemplate <- getTemplate templates "vouch.html"
+  return $ \UserFeature{userNameInPath, lookupUserName, lookupUserInfo, guardAuthenticated}
+            UploadFeature{uploadersGroup} -> do
+    let
+      handleGetVouches :: DynamicPath -> ServerPartE Response
+      handleGetVouches dpath = do
+        uid <- lookupUserName =<< userNameInPath dpath
+        userIds <- queryState vouchState $ GetVouchesFor uid
+        param <- renderToLBS lookupUserInfo userIds
+        pure . toResponse $ vouchTemplate
+          [ "msg" $= ""
+          , param
+          ]
+      handlePostVouch :: DynamicPath -> ServerPartE Response
+      handlePostVouch dpath = do
+        voucher <- guardAuthenticated
+        ugroup <- liftIO $ Group.queryUserGroup uploadersGroup
+        now <- liftIO getCurrentTime
+        vouchee <- lookupUserName =<< userNameInPath dpath
+        vouchersForVoucher <- queryState vouchState $ GetVouchesFor voucher
+        existingVouchers <- queryState vouchState $ GetVouchesFor vouchee
+        case join $ judge ugroup now vouchee vouchersForVoucher existingVouchers voucher of
```

Comment:
Why pawn off all checks to a judge function to immediatly unwrap them again?

Reply:
This is primarily such that the logic can be tested. I added tests in the latest comment.